pkg/customresourcestate/registry_factory.go

```diff
@@ -0,0 +1,517 @@
+						for _, v := range s {
+							if m, ok := v.(map[string]interface{}); ok {
+								candidate, set := m[key]
+								if !set {
+									continue
+								}
+
+								if candidate == val {
+									return m
+								}
+
+								if notNum == nil {
+									if i, err := getNum(candidate); err == nil && num == i {
+										return m
+									}
+								}
+
+								if notBool == nil {
+									if v, ok := candidate.(bool); ok && v == boolVal {
+										return m
+									}
+								}
+
+							}
+						}
+					}
+					return nil
+				},
+			})
+		} else {
+			out = append(out, pathOp{
+				part: part,
+				op: func(m interface{}) interface{} {
+					if mp, ok := m.(map[string]interface{}); ok {
+						return mp[part]
+					} else if s, ok := m.([]interface{}); ok {
+						i, err := strconv.Atoi(part)
+						if err != nil {
+							return nil
+						}
+						if i < 0 {
+							i += len(s)
+						}
+						if !(0 <= i && i < len(s)) {
+							return nil
+						}
+						return s[i]
+					}
+					return nil
+				},
+			})
+		}
+	}
+	return out, nil
+}
+
+func (s fieldMetrics) MetricFamilyGenerators(_, _ []string) (result []generator.FamilyGenerator) {
+	klog.Infof("custom resource state adding metrics: %v", s.names())
+	for _, f := range s.Families {
+		result = append(result, famGen(f))
+	}
+
+	return result
+}
+
+func famGen(f compiledFamily) generator.FamilyGenerator {
+	errLog := klog.V(f.ErrorLogV)
+	return generator.FamilyGenerator{
+		Name: f.Name,
+		Type: metric.Gauge,
+		Help: f.Help,
+		GenerateFunc: func(obj interface{}) *metric.Family {
+			return generate(obj.(*unstructured.Unstructured), f, errLog)
+		},
+	}
+}
+
+func generate(u *unstructured.Unstructured, f compiledFamily, errLog klog.Verbose) *metric.Family {
+	klog.V(10).Infof("%s: checking %s", f.Name, u.GetName())
+	var metrics []*metric.Metric
+	baseLabels := f.BaseLabels(u.Object)
+	values, errors := f.Each.Values(u.Object)
+
+	for _, err := range errors {
+		errLog.ErrorS(err, f.Name)
+	}
+
+	for _, v := range values {
+		v.DefaultLabels(baseLabels)
+		metrics = append(metrics, v.ToMetric())
+	}
+	klog.V(10).Infof("%s: produced %d metrics for %s", f.Name, len(metrics), u.GetName())
+
+	return &metric.Family{
+		Metrics: metrics,
+	}
+}
+
+func (s fieldMetrics) names() (names []string) {
+	for _, family := range s.Families {
+		names = append(names, family.Name)
+	}
+	return names
+}
+
+func getNum(value interface{}) (float64, error) {
+	var v float64
+	if value == nil {
+		return 0, fmt.Errorf("expected number but found nil value")
+	}
+	switch vv := value.(type) {
+	case bool:
+		if vv {
+			return 1, nil
+		}
+		return 0, nil
+	case string:
+		return strconv.ParseFloat(value.(string), 64)
+	case byte:
+		v = float64(vv)
+	case int:
+		v = float64(vv)
+	case int32:
+		v = float64(vv)
+	case int64:
+		v = float64(vv)
+	case uint:
+		v = float64(vv)
+	case uint32:
+		v = float64(vv)
+	case uint64:
+		v = float64(vv)
+	case float32:
+		v = float64(vv)
```

Comment:
Can these cases be merged into one like this?

```suggestion
	case byte, int, int32, int64, uint, uint32, uint64, float32:
		v = float64(vv)
```

Reply:
This did not work as a multi-type case results in an interface variable that I can't cast to flaot64.

Reply:
Ok, thanks for giving it a shot!

Reply:
Can't this be shortened with `default:`? Or has this the same drawback as the multi-type case?